aim/ext/cleanup/__init__.py

```diff
@@ -0,0 +1,91 @@
+from abc import abstractmethod
+
+import atexit
+import weakref
+import logging
+import threading
+
+from typing import Generic, Tuple, TypeVar
+
+T = TypeVar('T')
+
+logger = logging.getLogger(__name__)
+
+
+class RobustExec(threading.Thread):
+    """
+    A thread that executes a function and catches all soft interrupts.
+    Users very often Ctrl-C to stop the program multiple times which leaves
+    program no chance to clean up.
+    """
+    def join(self):
+        while True:
+            try:
+                # Pressing Ctrl-C will raise KeyboardInterrupt and stop only
+                # the `Thread.join()` call. The target function of `RobustExec`
+                # will continue to run.
+                return super().join()
+            except KeyboardInterrupt:
+                logger.warning("Received Ctrl-C. Closing gracefully.")
```

Comment:
@gorarakelyan This fully blocks force-Ctrl-C-aborting.
We can use this implementation now, but we should think of a way of letting user to abort the running process immediately.

Reply:
What about displaying prompt message asking whether user wants to terminate the process immediately?